src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -85,16 +89,17 @@ internal Solution(Workspace workspace, SolutionInfo.SolutionAttributes solutionA
         /// <summary>
         /// True if the solution contains a project with the specified project ID.
         /// </summary>
-        public bool ContainsProject(ProjectId projectId) => _state.ContainsProject(projectId);
+        public bool ContainsProject([NotNullWhen(returnValue: true)] ProjectId? projectId) => _state.ContainsProject(projectId);
 
         /// <summary>
         /// Gets the project in this solution with the specified project ID. 
         /// 
         /// If the id is not an id of a project that is part of this solution the method returns null.
         /// </summary>
-        public Project GetProject(ProjectId projectId)
+        // TODO: is it possible to annotate this method to say that if it returned non-null, then
+        // the input was non-null?
```

Comment:
@agocke @333fred I wasn't able to figure out how to annotate this. I want to say "if the return value is non-null, then the input was non-null", just like I can say `[NotNullWhen(returnValue: true)]`. There is `NotNullIfNotNullAttribute` which states "Specifies that the output will be non-null if the named parameter is non-null." but I want the reverse.

It's possible you don't support this due to complexity and I then get it, but it's very close to the ContainsProject case above that I wasn't sure how that's implemented. Clearly there's _some_ logic that can say "if the return value was checked, then we can infer information about the inputs" but maybe that's just restricted to Booleans because it fits nicely into if () checks.

Reply:
Yeah, I don't think we have an attribute that can support this currently. /cc @MadsTorgersen

Reply:
@jasonmalinowski I think you may be misunderstanding the attributes regardless. Split-state tracking only works through direct use of a conditional operation, so even if that attribute existed, your code would have to look something like

```C#
if (GetProject(projectId) != null)
{
   // projectId is non-null here
}
```

and all of the code that I can find that uses this method does something like

```C#
var project = GetProject(projectId);
if (project != null)
{
   // the test of the local variable has no effect on projectId
}
```

The same is true of any of the attributes or null checks.

Reply:
> Split-state tracking only works through direct use of a conditional operation

Yeah, makes sense; I just was really hoping you did fancier things. ;-) I also wasn't sure if due to lowering passes in the compiler whether the former case and the latter case you had were really one in the same.